### DISCUSSION TITLE
Invalidate does not work for complex query keys with nested objects.

### DISCUSSION THREAD

**Discussion by lisherwin** [TanStack/query#4891] [Q&A] [answered]
Hi! I am noticing that my calls to `invalidate()` do not properly invalidate and as a result `queryFn()` is not called for complex keys. In our app, we have fairly complex query keys. The query keys can be list of strings and objects, and those objects themselves can in turn have lists of objects. 

For example, we have query keys that look like:
```
const query = useQuery({...
    queryKey: [
        "friend",
        { limbs: [{ arms: 2, legs: 2 }] }
    ],
    queryFn: () => {...}
    ...
```
but when we call
```
queryClient.invalidateQueries([
    "friend",
    { limbs: [{ arms: 2, legs: 2 }] }
]);
```
`query`'s queryFn doesn't execute again. 
Here is the CodeSandbox: https://codesandbox.io/s/reverent-lumiere-movvbg?file=/src/App.tsx

From my understanding after reading https://stackoverflow.com/questions/73295707/how-does-reactquery-monitor-complex-query-key-changes-by-value-and-or-reference, it seems like React Query should be doing sorting and doing deep equals on the objects, and re-running the queryFn each time we invalidate. Am I missing something? Thank you in advance for any help or thoughts!

**Comment by louis-young** [comment, marked as answer]
Hi 👋 

Thanks for providing a reproduction. It's important to note that the reproduction doesn't model the example in your question.

You're creating the query key using this array:

```ts
const someObject = ["friend", { limbs: [{ arms: 2, legs: 2 }] }];

["friend", someObject]
```

And you're attempting to invalidate using `someObject` (or to be precise, the same array which is not referentially equal but that doesn't matter as keys are serialised).

This means you're saying "my query key starts with this":

```ts
["friend", { limbs: [{ arms: 2, legs: 2 }] }]
```

When it actually doesn't, it looks this:

```ts
["friend", ["friend", { limbs: [ { arms: 2, legs: 2 } ]} ]]
```

You probably want to call `invalidateQueries` with either of these arguments:

```ts
["friend", ["friend", { limbs: [ { arms: 2, legs: 2 } ]} ]]

["friend"]
```

Here's a forked fixed example: https://codesandbox.io/s/zealous-kate-ri2wq5?file=/src/App.tsx

Hopefully this helps 🙂

**Reply by lisherwin** [reply]
Thank you Louis! No you're totally right. My repro had a typo and this worked.
